### ISSUE TITLE
Simulate Geometric Brownian Motion

### ISSUE THREAD

**Issue by def-mycroft** [rcalxrc08/FinancialMonteCarlo.jl#4]
Hello, thanks in advance for any help. 

I want to generate GBM paths with a non-zero initial condition. 

I'm successful with this code: 
```julia
Nsim = Int(1e5)
Nstep = 250
σ = 0.2
μ = 0
mc = MonteCarloConfiguration(Nsim, Nstep)
rfCurve = ZeroRate([0.00, 0.02], 1.0);
X = simulate(BrownianMotion(σ, μ), mc, 1.0);

@assert (std(X[:, size(X, 2)]) - σ) / σ < 0.01
@assert abs(mean(X[:, size(X, 2)]) - μ) < 0.01
```

...but when I run this: 

```julia
simulate(GeometricBrownianMotion(σ, μ, 100), mc, 1.0)
```

I receive an error: 

```
MethodError: no method matching simulate(::GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ::MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, ::Float64)
```

I also tried: 

```julia
simulate(GeometricBrownianMotion(σ, μ, 100), rfCurve, mc, 1.0)
```

...in this case, I receive error: 

```
type GeometricBrownianMotion has no field underlying

Stacktrace:
  [1] getproperty(x::GeometricBrownianMotion{Float64, Int64, Int64, Float64}, f::Symbol)
    @ Base ./Base.jl:33
  [2] predict_output_type_zero(x::GeometricBrownianMotion{Float64, Int64, Int64, Float64})
    @ FinancialMonteCarlo ~/.julia/packages/FinancialMonteCarlo/CKQ9D/src/output_type.jl:43
  [3] (::FinancialMonteCarlo.var"#123#124")(y::GeometricBrownianMotion{Float64, Int64, Int64, Float64})
    @ FinancialMonteCarlo ~/.julia/packages/FinancialMonteCarlo/CKQ9D/src/output_type.jl:3
  [4] MappingRF
    @ ./reduce.jl:93 [inlined]
  [5] afoldl(::Base.MappingRF{FinancialMonteCarlo.var"#123#124", Base.BottomRF{typeof(Base.add_sum)}}, ::Base._InitialValue, ::GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ::ZeroRateCurve{Float64, Float64}, ::MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, ::Float64)
    @ Base ./operators.jl:533
  [6] _foldl_impl(op::Base.MappingRF{FinancialMonteCarlo.var"#123#124", Base.BottomRF{typeof(Base.add_sum)}}, init::Base._InitialValue, itr::Tuple{GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ZeroRateCurve{Float64, Float64}, MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, Float64})
    @ Base ./tuple.jl:263
  [7] foldl_impl(op::Base.MappingRF{FinancialMonteCarlo.var"#123#124", Base.BottomRF{typeof(Base.add_sum)}}, nt::Base._InitialValue, itr::Tuple{GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ZeroRateCurve{Float64, Float64}, MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, Float64})
    @ Base ./reduce.jl:48
  [8] mapfoldl_impl(f::FinancialMonteCarlo.var"#123#124", op::typeof(Base.add_sum), nt::Base._InitialValue, itr::Tuple{GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ZeroRateCurve{Float64, Float64}, MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, Float64})
    @ Base ./reduce.jl:44
  [9] mapfoldl(f::Function, op::Function, itr::Tuple{GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ZeroRateCurve{Float64, Float64}, MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, Float64}; init::Base._InitialValue)
    @ Base ./reduce.jl:160
 [10] mapfoldl
    @ ./reduce.jl:160 [inlined]
 [11] #mapreduce#218
    @ ./reduce.jl:287 [inlined]
 [12] mapreduce
    @ ./reduce.jl:287 [inlined]
 [13] #sum#221
    @ ./reduce.jl:501 [inlined]
 [14] sum(f::Function, a::Tuple{GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ZeroRateCurve{Float64, Float64}, MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, Float64})
    @ Base ./reduce.jl:501
 [15] predict_output_type_zero(::GeometricBrownianMotion{Float64, Int64, Int64, Float64}, ::ZeroRateCurve{Float64, Float64}, ::Vararg{Any, N} where N)
    @ FinancialMonteCarlo ~/.julia/packages/FinancialMonteCarlo/CKQ9D/src/output_type.jl:3
 [16] simulate(mcProcess::GeometricBrownianMotion{Float64, Int64, Int64, Float64}, zeroCurve::ZeroRateCurve{Float64, Float64}, mcBaseData::MonteCarloConfiguration{Int64, Int64, FinancialMonteCarlo.StandardMC, FinancialMonteCarlo.SerialMode{Random.MersenneTwister}}, T::Float64)
    @ FinancialMonteCarlo ~/.julia/packages/FinancialMonteCarlo/CKQ9D/src/models/base_models.jl:52
 [17] top-level scope
    @ In[71]:1
 [18] eval
    @ ./boot.jl:360 [inlined]
 [19] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base ./loading.jl:1116
```

**Comment by rcalxrc08** [comment]
Thank you very much for raising the issue.
There are two issues actually:
1)The first I think it is related to the fact that you are passing an integer as initial condition. If you switch to 100.0 it should work (but it will not because of 2).
I will implement a promotion in case of integers, but I strongly suggest you to switch to float where needed.
2)I haven't exposed the actual simulate function for Geometric Brownian Motion, you should use the method simulate! where you pass as first argument the matrix where you want the simulation. I will expose that function as well.

**Comment by def-mycroft** [comment]
nice! I thought there might be a relatively simple change. 

This works for me now:

```julia
Nsim = Int(1e5)
Nstep = 250
S0 = 100.0
σ = 0.2
μ = 0
mc = MonteCarloConfiguration(Nsim, Nstep)
X = Array{Float64}(undef, Nsim, Nstep+1)
FinancialMonteCarlo.simulate!(X, GeometricBrownianMotion(σ, μ, S0), mc, 1.0)

rets = (X[:, 1] .- X[:, size(X, 2)]) ./ X[:, 1]
@assert abs(std(rets) - σ) < 0.01
@assert abs(mean(rets)) / S0 < 0.01
```

**Comment by def-mycroft** [comment]
I really appreciate this project btw, I've been using QuantLib (python version), I see some relative advantages here :)

**Comment by rcalxrc08** [comment]
Thank you very much!

**Comment by rcalxrc08** [comment]
Just commited the fix. 
To be noticed that you shouldn't provide a zerorate object in the call to simulate for gbm.
The drift is carried by the parameters of the model.

**Comment by rcalxrc08** [comment]
Solved in v0.2.2, thanks!
